### PR TITLE
Test PaSta + Synchronicity, equality of different network constructions

### PR DESCRIPTION
**Pull request by joba00002**
### Prerequisites

- [x] I adhere to the coding conventions (described [here](../CONTRIBUTING.md)) in my code.
- [x] I have updated the copyright headers of the files I have modified.
- [x] I have written appropriate commit messages, i.e., I have recorded the goal, the need, the needed changes, and the location of my code modifications for each commit. This includes also, e.g., referencing to relevant issues.
- [x] I have put signed-off tags in *all* commits.
- [x] I have updated the changelog file [NEWS.md](../NEWS.md) appropriately.
- [x] I have checked whether I need to adjust the showcase file `showcase.R` with respect to my changes.
- [x] The pull request is opened against the branch `dev`.

### Description

Addresses #86

### CODE REVIEW

**Comment by joba00002** [comment]
As it turns out, despite three existing tests asserting just that, `split.networks.time.based` and splitting a multinetwork and extracting the author and bipartite network out of that do not generally result in the same networks.

This can be seen in the sample data by choosing author.relation = 'cochange' and artifact.relation = 'mail'.

The difference is in how isolated nodes are treated.
I suspect that the existing tests only worked because with our sample data, there were no isolated nodes.

**Comment by joba00002** [comment]
It might still make sense to assert that networks are equal when setting `remove.isolates = FALSE`. Unfortunately, this currently does not work as `extract.bipartite.network` has no such option.

**Comment by joba00002** [comment]
> It might still make sense to assert that networks are equal when setting `remove.isolates = FALSE`. Unfortunately, this currently does not work as `extract.bipartite.network` has no such option.

I have added this option and changed the tests to use it.

**Comment by bockthom** [comment]
> As it turns out, despite three existing tests asserting just that, `split.networks.time.based` and splitting a multinetwork and extracting the author and bipartite network out of that do not generally result in the same networks.
> 
> This can be seen in the sample data by choosing author.relation = 'cochange' and artifact.relation = 'mail'.
> 
> The difference is in how isolated nodes are treated. I suspect that the existing tests only worked because with our sample data, there were no isolated nodes.

First of all, sorry for the late reply. Our mail servers have not been available and are (still!) not available. So, I might not directly notice when there are new postings.

Thanks for spotting all these inconsistencies and bugs, also thanks for adding a `remove.isolates` parameter for bipartite networks, sounds good!

I will probably have a look at your changes today or tomorrow, but can you please briefly explain what exactly the difference is in how "isolated nodes are treated"? Knowing that would be helpful for understanding and reviewing your changes.

**Comment by joba00002** [comment]
The difference is whether, or rather when, isolated nodes are removed.

Removing isolated nodes first and then splitting may lead to a different result than splitting first and then removing isolated nodes.

**Review by hechtlC — APPROVED** [review]
Looks good to me thank you @joba00002! I have just identified a couple small things that can be addressed quickly and after you have done that I am fine with merging this. @bockthom will also review so maybe wait with these changes until you have the notes from him.

File: tests/test-split-data-time-based.R [hunk 1]
```diff
@@ -443,9 +483,9 @@ test_that("Split a data object time-based (split.basis = 'commits', sliding.wind
         mails = list(
             "2016-07-12 15:58:59-2016-07-12 16:01:59" = data$mails[0, ],
             "2016-07-12 16:00:29-2016-07-12 16:03:29" = data$mails[0, ],
-            "2016-07-12 16:01:59-2016-07-12 16:04:59" = data$mails[rownames(data$mails) == 16, ],
-            "2016-07-12 16:03:29-2016-07-12 16:06:29" = data$mails[rownames(data$mails) %in% c(16, 17), ],
-            "2016-07-12 16:04:59-2016-07-12 16:06:33" = data$mails[rownames(data$mails) == 17, ]
+            "2016-07-12 16:01:59-2016-07-12 16:04:59" = data$mails[15, ],
+            "2016-07-12 16:03:29-2016-07-12 16:06:29" = data$mails[15:16, ],
+            "2016-07-12 16:04:59-2016-07-12 16:06:33" = data$mails[16, ]
```
Comment (bockthom) [inline tests/test-split-data-time-based.R:488]: Same as above.

File: tests/test-split-data-time-based.R [hunk 2]
```diff
@@ -470,20 +510,33 @@ test_that("Split a data object time-based (split.basis = 'commits', sliding.wind
...
         synchronicity = lapply(results, function(cf.data) cf.data$get.synchronicity())
     )
+    expected.data = lapply(expected.data, function(d) lapply(d, function (df) {
+        row.names(df) = NULL
+        return(df)
+    }))
+    results.data = lapply(results.data, function(d) lapply(d, function (df) {
+        row.names(df) = NULL
+        return(df)
+    }))
```
Comment (bockthom) [inline tests/test-split-data-time-based.R:520]: Same as above.

File: tests/test-split-data-time-based.R [hunk 3]
```diff
@@ -568,13 +621,13 @@ test_that("Split a data object time-based (split.basis = 'mails', sliding.window
...
+            "2004-10-09 18:38:13-2007-10-10 12:38:13" = data$mails[1:2, ],
             "2006-04-10 15:38:13-2009-04-10 09:38:13" = data$mails[0, ],
-            "2007-10-10 12:38:13-2010-10-10 06:38:13" = data$mails[rownames(data$mails) %in% 3:12, ],
-            "2009-04-10 09:38:13-2012-04-10 03:38:13" = data$mails[rownames(data$mails) %in% 3:12, ],
+            "2007-10-10 12:38:13-2010-10-10 06:38:13" = data$mails[3:12, ],
+            "2009-04-10 09:38:13-2012-04-10 03:38:13" = data$mails[3:12, ],
             "2010-10-10 06:38:13-2013-10-10 00:38:13" = data$mails[0, ],
             "2012-04-10 03:38:13-2015-04-10 21:38:13" = data$mails[0, ],
-            "2013-10-10 00:38:13-2016-07-12 16:05:38" = data$mails[rownames(data$mails) %in% 13:17, ]
+            "2013-10-10 00:38:13-2016-07-12 16:05:38" = data$mails[13:16, ]
```
Comment (bockthom) [inline tests/test-split-data-time-based.R:630]: Same as above.

File: tests/test-split-data-time-based.R [hunk 4]
```diff
@@ -603,20 +656,33 @@ test_that("Split a data object time-based (split.basis = 'mails', sliding.window
...
         synchronicity = lapply(results, function(cf.data) cf.data$get.synchronicity())
     )
+    expected.data = lapply(expected.data, function(d) lapply(d, function (df) {
+        row.names(df) = NULL
+        return(df)
+    }))
+    results.data = lapply(results.data, function(d) lapply(d, function (df) {
+        row.names(df) = NULL
+        return(df)
+    }))
```
Comment (bockthom) [inline tests/test-split-data-time-based.R:666]: Same as above.

File: tests/test-split-data-time-based.R [hunk 5]
```diff
@@ -689,8 +755,8 @@ test_that("Split a data object time-based (split.basis = 'issues', sliding.windo
         mails = list(
             "2013-04-21 23:52:09-2015-04-22 11:52:09" = data$mails[0, ],
             "2014-04-22 05:52:09-2016-04-21 17:52:09" = data$mails[0, ],
-            "2015-04-22 11:52:09-2017-04-21 23:52:09" = data$mails[rownames(data$mails) %in% 14:17, ],
-            "2016-04-21 17:52:09-2017-05-23 12:32:40" = data$mails[rownames(data$mails) %in% 14:17, ]
+            "2015-04-22 11:52:09-2017-04-21 23:52:09" = data$mails[13:16, ],
+            "2016-04-21 17:52:09-2017-05-23 12:32:40" = data$mails[13:16, ]
```
Comment (bockthom) [inline tests/test-split-data-time-based.R:759]: Same as above.

File: tests/test-split-data-time-based.R [hunk 6]
```diff
@@ -713,21 +779,34 @@ test_that("Split a data object time-based (split.basis = 'issues', sliding.windo
...
         synchronicity = lapply(results, function(cf.data) cf.data$get.synchronicity())
     )
+    expected.data = lapply(expected.data, function(d) lapply(d, function (df) {
+        row.names(df) = NULL
+        return(df)
+    }))
+    results.data = lapply(results.data, function(d) lapply(d, function (df) {
+        row.names(df) = NULL
+        return(df)
+    }))
```
Comment (bockthom) [inline tests/test-split-data-time-based.R:789]: Same as above.

File: tests/test-split-data-time-based.R [hunk 7]
```diff
@@ -1300,7 +1448,7 @@ test_that("Split a data object time-based with equal-sized windows (number.windo
         mails = list(
             "2013-04-21 23:52:09-2014-09-01 12:05:39" = data$mails[0, ],
             "2014-09-01 12:05:39-2016-01-12 00:19:09" = data$mails[0, ],
-            "2016-01-12 00:19:09-2017-05-23 12:32:40" = data$mails[rownames(data$mails) %in% 14:17, ]
+            "2016-01-12 00:19:09-2017-05-23 12:32:40" = data$mails[13:16, ]
```
Comment (bockthom) [inline tests/test-split-data-time-based.R:1451]: Same as above.

**Review by bockthom — COMMENTED** [review]
Looks also good to me @joba00002.

I have only two aesthetic comments:
- Removing the row names using two nested lapply calls occurs multiple times in the adjusted tests. Might be worth to extract a helper function for that, as this would increase the readability and understandability of the tests.
- In the tests, where you needed to adjust the expected data (as row names have changed), I would like to keep the previous row name as a comment at the end of the line, as the previously used row names map to the row names in the *.list files (if I remember correctly), which would ease manual debugging with the test data.

Please see my comments below, most of them deal with the two aesthetic problems mentioned above.

File: tests/test-split-data-time-based.R [hunk 8]
```diff
@@ -141,20 +142,33 @@ test_that("Split a data object time-based (split.basis = 'commits').", {
...
     )
 
+    expected.data = lapply(expected.data, function(d) lapply(d, function (df) {
+        row.names(df) = NULL
+        return(df)
+    }))
+    results.data = lapply(results.data, function(d) lapply(d, function (df) {
+        row.names(df) = NULL
+        return(df)
+    }))
```
[the remark this answers is not in the document]
Comment (joba00002) [reply tests/test-split-data-time-based.R]: I thought the same thing, so `testing-utils.R` is where they are for now.
If we find a use for them outside of testing, it is easy to move them to `util-misc.R`.

I'm also not completely happy with the names, as they are very long. I couldn't think of anything shorter that describes what they do, though.

File: tests/testing-utils.R [hunk 9]
```diff
@@ -87,3 +87,29 @@ cases.cross.product = function (cases.1, cases.2) {
...
+        return(df)
+    })
+}
+
+#' Remove the row names in a list of range data objects, which are lists of dataframes.
+#' @seealso remove.row.names.from.data
+#'
+#' @param list.of.range.data.objects a list of range data objects
+#'
+#' @returns the list of range data objects, but without row names
```
Comment (bockthom) [inline tests/testing-utils.R:112]: To be more precise: ...but without row names for their dataframes

**Review by bockthom — CHANGES_REQUESTED** [review]

File: tests/test-split-data-time-based.R (hunk 8 above)
[the remark this answers is not in the document]
Comment (bockthom) [reply tests/test-split-data-time-based.R]: It is always a big hassle to find proper function names that are precise and short. So, for me, the names are ok. @hechtlC What's your opinion on that?

[the remark this answers is not in the document]
Comment (hechtlC) [reply tests/test-split-data-time-based.R]: How about `remove.row.names.from.inner.list.of.dfs`?

File: tests/testing-utils.R [hunk 10]
```diff
@@ -87,3 +87,31 @@ cases.cross.product = function (cases.1, cases.2) {
...
+
+#' Remove the row names of each data frame in a list of dataframes.
+#' Useful for comparing data where row names are allowed to change.
+#'
+#' Note that removing row names causes them to be set to increasing numbers starting at one,
+#' so this function may have to be called on both lists involved in a comparison.
+#'
+#' @param data a list of dataframes
+#'
+#' @returns the list of dataframes, but without row names
```
Comment (hechtlC) [inline tests/testing-utils.R:99]: Please remove the s from returns.
Same for the next function.

File: tests/testing-utils.R [hunk 11]
```diff
@@ -106,12 +106,12 @@ remove.row.names.from.data = function(data) {
     )
 }
 
-#' Remove the row names in a list of range data objects, which are lists of dataframes.
+#' Remove the row names in a list of lists of dataframes, for example a list of range data objects.
```
Comment (bockthom) [inline tests/testing-utils.R:109]: Could you please describe more precise from where the names are removed:
"Remove the row names from the dataframes in a list of ...."

File: tests/testing-utils.R [hunk 12]
```diff
@@ -106,12 +106,12 @@ remove.row.names.from.data = function(data) {
...
 #' @seealso remove.row.names.from.data
 #'
-#' @param list.of.range.data.objects a list of range data objects
+#' @param list.of.lists.of.dfs a list of lists of dataframes
 #'
 #' @return the list of range data objects, but without row names for their data frames
```
Comment (bockthom) [inline tests/testing-utils.R:114]: The description of the return value needs to be updated as it still talks ab out "range data objects".

**Review by bockthom — COMMENTED** [review]
Almost done :wink: I just spotted two sentences in the function documentation of the function we have renamed. Please fix them and then I will merge right away.

**Review by bockthom — APPROVED** [review]
Looks good!

**Pull request merged**